cli/src/pcluster/models/common.py

```diff
@@ -239,3 +251,330 @@ def dump_json(self):
+
+    # --------------------------------------- S3 private functions --------------------------------------- #
+
+    def _upload_file(self, content, file_name, file_type, format=S3FileFormat.YAML):
+        """Upload file to S3 bucket."""
+        if format == S3FileFormat.YAML:
+            result = AWSApi.instance().s3.put_object(
+                bucket_name=self.name,
+                body=yaml.dump(content),
+                key=self.get_object_key(file_type, file_name),
+            )
+        elif format == S3FileFormat.JSON:
+            result = AWSApi.instance().s3.put_object(
+                bucket_name=self.name,
+                body=json.dumps(content),
+                key=self.get_object_key(file_type, file_name),
+            )
+        else:
+            result = AWSApi.instance().s3.put_object(
+                bucket_name=self.name,
+                body=content,
+                key=self.get_object_key(file_type, file_name),
+            )
+        return result
+
+    def _get_file(self, file_name, file_type, version_id=None):
+        """Get file from S3 bucket."""
+        result = AWSApi.instance().s3.get_object(
+            bucket_name=self.name, key=self.get_object_key(file_type, file_name), version_id=version_id
+        )
+
+        config_content = result["Body"].read().decode("utf-8")
+        return yaml.safe_load(config_content)
+
+    def _get_file_url(self, file_name, file_type):
+        """Get file url from S3 bucket."""
+        url = "https://{bucket_name}.s3.{region}.amazonaws.com{partition_suffix}/{config_key}".format(
+            bucket_name=self.name,
+            region=self.region,
+            partition_suffix=".cn" if self.region.startswith("cn") else "",
+            config_key=self.get_object_key(file_type, file_name),
+        )
+        return url
+
+
+class S3BucketFactory:
+    """S3 bucket factory to return a bucket object with existence check and creation."""
+
+    @staticmethod
+    def init_s3_bucket(service_name, artifact_directory, stack_name, custom_s3_bucket):  # noqa: C901
+        """Initialize the s3 bucket."""
+        if custom_s3_bucket:
+            bucket = S3Bucket(
+                service_name=service_name,
+                name=custom_s3_bucket,
+                artifact_directory=artifact_directory,
+                stack_name=stack_name,
+                is_custom_bucket=True,
+            )
+            try:
+                bucket.check_bucket_exists()
+            except AWSClientError as e:
+                raise AWSClientError(
+                    "check_bucket_exists", f"Unable to access config-specified S3 bucket {bucket.name}. Due to {str(e)}"
+                )
+        else:
+            bucket = S3Bucket(service_name=service_name, artifact_directory=artifact_directory, stack_name=stack_name)
+            try:
+                bucket.check_bucket_exists()
```

Comment:
I don't think we can use head_object failure to determine the bucket existence, because the error message for head_object if bucket doesn't exist is:
```
Error during execution of head_object. Unexpected error when calling get_bucket_location on S3 bucket 'hello': 'Forbidden'. Error code: 403
```
it could be permission deny.
Instead, head_bucket gives error `Not Found` is clearer.

Reply:
ok then let's leave it as is. I wish we could spare the extra call.